### PR TITLE
[Bugfix] Fix full-payload accumulation across preemption and aborts

### LINKED ISSUE #7569: [Bug][Lifecycle] Shared AR→DiT payload accumulator: preemption/resume duplication and flush-on-abort
## Summary

Two behaviours in the shared cross-step payload accumulator were identified
while reviewing the MammothModa2 prefix-cache and AR→DiT payload work
(#7102, #7119, #7134, #7476). Both are potential lifecycle risks in shared
code. Neither is introduced by a specific model PR, and neither currently has
a minimal reproduction.

**Reachability.** The accumulator is gated by
`should_accumulate_full_payload_output`
(`vllm_omni/distributed/omni_connectors/model_runner/omni_connector_runtime.py:59`),
which fires only for a stage that declares `custom_process_next_stage_input_func`,
is not `final_output`, and is not `async_chunk`. That is 15 pipelines today:
audex, aura_omni, breeze_tts_2, cosyvoice3, covo_audio, dynin_omni, indextts2,
joyai_vl_interaction, mimo_audio, minicpmo_4_5, nemotron_voicechat, personaplex,
qwen2_5_omni, qwen3_omni, qwen3_tts.

MammothModa2 does **not** reach this path: stage 0 declares neither hook, and
stage 1 sets `final_output=True`. Upstream pins this in
`tests/worker/test_omni_connector_mixin.py:397`, which lists
`mammoth_moda2.ar2dit` among the `full_payload_mode` incompatible paths. The
MammothModa2 PRs above are listed only as the review origin, not as affected
consumers.

### 1. Append-only accumulation has no positional reconciliation

`accumulate_full_payload_output`
(`vllm_omni/distributed/omni_connectors/model_runner/omni_connector_payload_transport.py:611`)
appends per-step payloads into `_pending_full_payload_send[req_id]`
(`chunks[k].append(v)` at `:664`) and materialises them with a plain
`torch.cat` (`:556`). There is no logical-position key and no replacement
operation.

If a preempted-and-resumed request re-emits rows that were already emitted
before preemption, those rows would be appended a second time rather than
replacing the earlier rows, so the accumulated payload length could exceed the
logical sequence length expected downstream.

The append/cat behaviour is confirmed by source inspection. The
scheduler-driven preemption/resume trigger has not been reproduced.

### 2. `cleanup_finished_request` flushes pending data

`cleanup_finished_request`
(`vllm_omni/distributed/omni_connectors/model_runner/omni_connector_runtime.py:313`)
calls `flush_full_payload_outputs({req_id})` at `:339` when a request still has
a pending accumulator entry. If an aborted request reaches this path, a partial
payload may be sent downstream rather than discarded.

There is no lifecycle-level guard for this. A downstream length assertion is a
correctness check, not a cleanup mechanism, and assertions are removed under
`python -O`.

## Scope

This is shared connector/request lifecycle code. A fix should not be bundled
into a model-specific PR. Establish a minimal reproduction and lifecycle tests
before changing the shared implementation.

The same class of risk was previously observed during the model-local GPU
accumulator exploration for #7102, but that does not by itself prove a
regression in this shared path.

## Requested work

1. Add a scheduler-driven preemption/resume test that verifies whether
   re-emitted logical rows are duplicated or reconciled.
2. Add an abort-during-generation test that verifies whether pending payloads
   are discarded or sent.
3. Define the lifecycle contract for pending payloads and request identity
   across finish, cancel, preemption, and retry.
4. Decide whether the accumulator must become position-aware, or whether the
   abort path should explicitly discard pending payloads while preserving
   normal completion flushing.
5. Validate the decision with mixed-request coverage on at least one affected
   pipeline.

## Acceptance criteria

- The expected behaviour for finish, cancel, preemption, and retry is documented.
- Tests cover both normal completion and abnormal termination.
- Request A cannot duplicate or leak rows into request B.
- A cancelled request cannot send a partial payload downstream.
- The fix, if required, is isolated to the shared lifecycle owner.

---

*Line references are to post-split locations; the connector model-runner mixin
was split out of `vllm_omni/worker/omni_connector_model_runner_mixin.py` by
#6903, which is now a 27-line re-export shim.*

### PR DESCRIPTION
**Pull request by Bezdarnost**
## Purpose

Fixes #7569.

Full-payload accumulation currently appends replayed AR rows after preemption and flushes partial output when a request is aborted. The preemption problem was reproduced with Qwen3-TTS: forcing the scheduler's normal preemption path produced 71 accumulated codec rows for a 39-token sequence.

Track each payload key's emitted input-token boundary in the shared accumulator. Preserve existing rows across preemption, skip replayed prefixes, and append only the new suffix of overlapping token-aligned output. Keep snapshot/replace and scalar metadata policies, with replayed older updates suppressed. Release position tracking on terminal cleanup so a fresh attempt can reuse the ID.

Carry aborted/errored/ignored request IDs from both schedulers to GPU/NPU runners. Both the pre-execution flush and cleanup fallback discard these pending payloads; normal completion still flushes. Document completion, failure, preemption and retry semantics in the connector module design.

Partially overlapping non-token-aligned tensors raise an explicit error because their rows cannot safely be mapped to token positions. One-shot generation payloads retain their existing policy. Already submitted connector writes are outside the pending accumulator's ownership.

## Test Plan

Focused regressions (no weights or GPU required):

```bash
python -m pytest tests/worker/test_full_payload_lifecycle.py -q
```

The eight cases selected by `-k 'test_execute_discards or test_update_states'` fail on unmodified base `c32aaeb2` because the cancelled request is sent downstream. New preemption coverage calls the real scheduler preemption method, replays differently sized chunks across the previous boundary, and checks exact row values and request isolation. Other coverage includes normal completion, abort/error/ignored statuses, waiting/running cancellation, abort-all, request-ID reuse, recovered prefixes, metadata policies and ambiguous replay rejection. An existing output-builder test checks the token ranges passed by the AR runner.

Broader CI-aligned suite:

```bash
python -m pytest tests/core/sched tests/worker tests/distributed/omni_connectors -m 'core_model and cpu' --run-level core_model -q -rs
```

The new module is selected by the existing `Simple · Other Test` step in `.buildkite/cuda/test-ready.yml`; no CI routing changes are needed.

**vLLM Version:** 0.29.0; Python 3.12.11, torch 2.13.0, Transformers 5.14.1.

**vLLM-Omni Commit:** base `c32aaeb2` plus this patch.

## Test Result

- **38 new lifecycle cases passed.** Broader suite: **921 passed, 1 skipped, 34 deselected**, plus three passing subtests. The skipped existing test has no connector config files at its expected location.
- Full-model validation: **Qwen/Qwen3-TTS-12Hz-1.7B-CustomVoice**, one **RTX 5080 16 GB**, full-payload mode (`async_chunk: false`), eager execution, async scheduling enabled. A local harness forced the real scheduler `_preempt_request` path after 12 generated tokens, then allowed normal scheduling/resumption. Before reconciliation: **71 rows / 39 tokens**. On the final patch, with two requests active: **44 rows / 45 tokens**, the expected one row per computed input token; both requests completed. These are separate stochastic runs, not an output-quality parity comparison.
- Two further disconnect cycles waited for worker-side accumulation before cancelling a long request. Neither cancelled payload entered `send_full_payload_outputs`; concurrent and subsequent requests completed. Seven completed responses, including warmup and the preemption pair, contained valid 24 kHz WAV audio. This was a manual instrumented run, not a new model-level CI test or a natural memory-pressure benchmark.
- Applicable pre-commit checks passed except mypy, which reports the same **100 existing diagnostics** on base and branch for the expanded file set: **zero new diagnostics** after normalizing line numbers. The new regression module passes mypy.
- NPU call sites were updated and statically checked; NPU hardware execution was unavailable. The installed vLLM-Omni version metadata emits a mismatch warning with vLLM 0.29.0; all results above were obtained in that environment.

### CODE REVIEW

**Comment by Bezdarnost** [comment]
> This PR appears to belong to: **docs/design/module/ar_runtime.md**, **docs/design/module/omni_connector.md**.
> 
> Module owners: @fake0fan @Gaohan123 @Sy0307
> 
> Routing: @fake0fan via module of the changed files, semantic router; @Gaohan123 via module of the changed files, CODEOWNERS; @Sy0307 via module of the changed files
> 
> @Bezdarnost, please review your own changes and leave a short self-review comment describing what you checked. PRs without author self-review may not be assigned a reviewer.
> 
> Please take a look when you have a chance. If you would like an automated review, mention @vllm-omni-review-bot in a comment.

Author self-review:

- Checked both issues in #7569: replayed AR prefixes no longer append duplicate rows, and cancellation/failure discards pending output in both flush paths.
- Reviewed position tracking and cleanup across completion, preemption, errors and fresh retries, including internal request-ID reuse. Added the lifecycle contract to the connector design.
- Ran 38 new lifecycle cases and the broader scheduler/worker/connector suite: 921 passed, 1 skipped, 34 deselected. Eight cancellation regressions fail on the unmodified base.
- Reproduced preemption on Qwen3-TTS with an RTX 5080: before reconciliation, 71 codec rows for 39 tokens; the final patch produced the expected 44 rows for 45 tokens while another request was active. Both requests completed. Two cancellation cycles also passed, with no cancelled payload sent and healthy concurrent/follow-up requests.
- Checked GPU/NPU call-site parity. NPU execution and perceptual audio-quality parity were not tested. Ambiguous partial replay of non-token-aligned tensors is explicitly rejected.
- Ran pre-commit. Mypy reports the same 100 existing diagnostics on base and branch, with zero new diagnostics; the new regression module passes.

File: vllm_omni/distributed/omni_connectors/model_runner/omni_connector_payload_transport.py [hunk 1]
```diff
@@ -608,18 +608,61 @@ def _resolve_full_payload_replace_keys(self) -> frozenset:
...
+            if value is None:
+                continue
+            previous_end = token_ends.get(key)
+            if previous_end is not None:
+                if end <= previous_end:
+                    continue
+                if isinstance(value, torch.Tensor) and value.ndim >= 2 and key not in replace_keys:
+                    num_rows = value.shape[0]
+                    # Prefix-cache recovery can include rows before this step.
+                    row_start = 0 if num_rows == end else start
```
[the remark this answers is not in the document]
Comment (Bezdarnost) [reply vllm_omni/distributed/omni_connectors/model_runner/omni_connector_payload_transport.py]: Replaced the row-count heuristic with explicit prefix-cache provenance from the GPU and NPU AR runners. Only cache-merged hidden states and emitted `mm_cache_keys` receive prefix handling; passthrough fields remain ordinary deltas.

Added regression coverage for a `[2, 4)` step emitting four audio rows: a non-overlapping delta retains every row, while an ambiguous partial replay raises the existing alignment error without changing accumulated data or positions. Added runner coverage verifying that passthrough and non-emitted keys are excluded from the cache metadata.

All four new regression cases fail on the previous PR head (`90eb9ee7`) and pass with this fix. Validation: 105 focused tests passed; the broader CPU suite passed 956 tests (one existing configuration-dependent skip). Formatting/lint checks passed; mypy diagnostics are unchanged from the previous head. This follow-up has not been run on NPU hardware or through another full-model GPU run.
